js/src/components/ShinyMarkdownStream.tsx

```diff
@@ -0,0 +1,244 @@
+    } catch (err) {
+      showShinyClientMessage({
+        status: "error",
+        message: `Failed to initialize Shiny inputs: ${err}`,
+      })
+    }
+
+    try {
+      await shiny.bindAll(el)
+    } catch (err) {
+      showShinyClientMessage({
+        status: "error",
+        message: `Failed to bind Shiny inputs/outputs: ${err}`,
+      })
+    }
+  }
+
+  @throttle(200)
+  private static async _throttledBind(el: HTMLElement): Promise<void> {
+    await this.#doBind(el)
+  }
+
+  static async #doUnBind(el: HTMLElement): Promise<void> {
+    const shiny = window?.Shiny as Shiny
+    if (!shiny?.unbindAll) return
+
+    try {
+      shiny.unbindAll(el)
+    } catch (err) {
+      showShinyClientMessage({
+        status: "error",
+        message: `Failed to unbind Shiny inputs/outputs: ${err}`,
+      })
+    }
+  }
+
+  connectedCallback() {
+    // Don't use shadow DOM so the component can inherit styles from the main document
+    const root = document.createElement("div")
+    root.classList.add("html-fill-container", "html-fill-item")
+    this.appendChild(root)
+
+    this.rootElement = root
+
+    // Read initial attributes
+    this.content = this.getAttribute("content") || ""
+    this.contentType =
+      (this.getAttribute("content-type") as ContentType) || "markdown"
+    this.streaming = this.hasAttribute("streaming")
+    this.autoScroll = this.hasAttribute("auto-scroll")
+    this.codeTheme = {
+      light: this.getAttribute("code-theme-light") || "atom-one-light",
+      dark: this.getAttribute("code-theme-dark") || "atom-one-dark",
+    }
+
+    // Initial render
+    this.renderValue()
+  }
+
+  disconnectedCallback() {
+    if (this.rootElement) {
+      render(null, this.rootElement)
+    }
+  }
+
+  private renderValue() {
+    if (!this.rootElement) return
+
+    render(
+      <StrictMode>
+        <MarkdownStream
+          content={this.content}
+          contentType={this.contentType}
+          streaming={this.streaming}
+          autoScroll={this.autoScroll}
+          onContentChange={this.handleContentChange}
+          onStreamEnd={this.handleStreamEnd}
+          onWillContentChange={this.handleWillContentChange}
+          codeThemeLight={this.codeTheme.light}
+          codeThemeDark={this.codeTheme.dark}
+        />
+      </StrictMode>,
+      this.rootElement,
+    )
+  }
+
+  handleWillContentChange = () => {
+    ShinyMarkdownStreamOutput.#doUnBind(this)
+  }
+
+  // Public callback methods for React component and testing
+  handleContentChange = () => {
+    // Render Shiny HTML dependencies and bind inputs/outputs
+    if (this.streaming) {
+      ShinyMarkdownStreamOutput._throttledBind(this)
+    } else {
+      ShinyMarkdownStreamOutput.#doBind(this)
+    }
+
+    // Callback for when content changes - can be used for custom logic
+    this.dispatchEvent(
+      new CustomEvent("contentchange", {
+        detail: { content: this.content },
+      }),
+    )
+  }
```

Comment:
Feels like the binding logic could be moved into the `MarkdownStream` component implementation?